### ISSUE TITLE
id(token) gives surprising results

### ISSUE THREAD

**Issue by f11r** [explosion/spaCy#1573]
It seems to me that the identity function of `Token` is either broken in v2 or I missed the documentation that it has changed between v1 and v2.

Running the code 
(v1.9)
~~~
import en_depent_web_md
nlp = en_depent_web_md.load()
~~~
(v2.0)
~~~
import en_core_web_sm
nlp = en_core_web_sm.load()
~~~

and then

~~~
doc = nlp('I am')
for t in doc:
    print(t, t.head, t is t.head, id(t), id(t.head))
~~~

gives

(v1.9)
~~~
I am False 140040072365864 140040072487544
am am True 140040072487544 140040072487544
~~~

(v2.0)
~~~
I am False 140185708576344 140185708574664
am am False 140185708575464 140185708574664
~~~

and the return value of `doc[0].head is doc[1]` is `True` in v1.9 and `False` in v2.0.

Sorry if I just missed this in the new documentation.

## Your Environment

* **spaCy version:** 1.9.0
* **Platform:** Linux-4.4.59-x86_64-with-fedora-26-Twenty_Six
* **Python version:** 3.6.2
* https://github.com/explosion/spacy-models/releases/download/en_depent_web_md-1.2.1/en_depent_web_md-1.2.1.tar.gz

and 

* **spaCy version:** 2.0.1
* **Platform:** Linux-4.4.59-x86_64-with-fedora-26-Twenty_Six
* **Python version:** 3.6.2
* https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-2.0.0/en_core_web_sm-2.0.0.tar.gz

**Comment by honnibal** [comment]
Object identity has never been reliable for tokens in spaCy, because the `Doc` object would create and destroy the `Token` objects as the `Doc` was updated. However, in v1 the identity *seemed* to work, because we cached the objects.

The caching wasn't actually useful for performance, and it creates a cycle that meant the `Doc` object couldn't be ref counted (it had to be garbage collected). So now we just create a new `Token` object every time we need one.

Depending on what you're trying to do, there could be a number of solutions. `(id(token.doc), token.idx)` would be one solution.
